rs/execution_environment/src/execution/update.rs

```diff
@@ -341,26 +359,22 @@ impl UpdateHelper {
 
         validate_message(&canister, &original.method)?;
 
-        if let CanisterCallOrTask::Call(_) = original.call_or_task {
-            // TODO(RUN-957): Enforce the limit in heartbeat and timer after
-            // canister logging ships by removing the `if` above.
+        let wasm_memory_usage = canister
+            .execution_state
+            .as_ref()
+            .map_or(NumBytes::new(0), |es| {
+                num_bytes_try_from(es.wasm_memory.size).unwrap()
+            });
 
-            let wasm_memory_usage = canister
-                .execution_state
-                .as_ref()
-                .map_or(NumBytes::new(0), |es| {
-                    num_bytes_try_from(es.wasm_memory.size).unwrap()
-                });
+        if let Some(wasm_memory_limit) = clean_canister.system_state.wasm_memory_limit {
+            // A Wasm memory limit of 0 means unlimited.
+            if wasm_memory_limit.get() != 0 && wasm_memory_usage > wasm_memory_limit {
+                let err = HypervisorError::WasmMemoryLimitExceeded {
+                    bytes: wasm_memory_usage,
+                    limit: wasm_memory_limit,
+                };
```

Comment:
@mraszyk Here we check for the error. So it is immediately discovered before starting execution.

Reply:
Unless I'm missing something, we're charging the base fee of 5M cycles per execution nonetheless.

Reply:
You may be right, that is a good point.

Reply:
I will close this thread, and let's continue in the one above.